runner/lib/helpers/buffer-line-transform.js

```diff
@@ -0,0 +1,95 @@
+/* global Buffer */
+/* eslint-disable no-underscore-dangle,no-nested-ternary,no-plusplus */
+
+import { Transform } from 'stream';
+
+export default class BufferLineTransform extends Transform {
+  /**
+   * The BufferLineTransform is reading String or Buffer content from a Readable stream
+   * and writing each line as a Buffer in object mode
+   *
+   * @param {import('./buffer-line-transform.js').BufferLineTransformOptions} [options]
+   */
+  constructor(options) {
+    const { break: breakValue, breakEncoding, ...transformOptions } =
+      options || {};
+    super({ ...transformOptions, readableObjectMode: true });
+    this._breakValue = breakValue || 10;
+    this._breakEncoding = breakEncoding;
+
+    /** @type {number} */
+    let breakLength;
+    if (!breakValue || typeof breakValue === 'number') {
+      breakLength = 1;
+    } else if (Buffer.isBuffer(breakValue)) {
+      breakLength = breakValue.length;
+    } else {
+      breakLength = Buffer.from(breakValue, breakEncoding).length;
+    }
+    this._breakLength = breakLength;
+
+    /** @type {Array<Buffer>} */
+    this._chunks = [];
+  }
+
+  /**
+   * @param {any} chunk
+   * @param {BufferEncoding | 'buffer'} encoding
+   * @param {import('stream').TransformCallback} cb
+   * */
+  _transform(chunk, encoding, cb) {
+    try {
+      /** @type {Buffer} */
+      let buf =
+        Buffer.isBuffer(chunk) || encoding === 'buffer'
+          ? chunk
+          : Buffer.from(chunk, encoding);
+
+      while (buf.length) {
+        const offset = buf.indexOf(this._breakValue, 0, this._breakEncoding);
+
+        /** @type {number} */
+        let endOffset;
+        if (offset >= 0) {
+          endOffset = offset + this._breakLength;
+          if (this._chunks.length) {
+            const concatLength = this._chunks.reduce(
+              (acc, { length }) => acc + length,
+              endOffset,
+            );
+            this._writeItem(
+              Buffer.concat(
+                [...this._chunks.splice(0, this._chunks.length), buf],
+                concatLength,
+              ),
```

Comment:
Buffer concatenation is expensive and avoidable.

Reply:
Unfortunately I didn't find a way to avoid it here. The stream is in block mode to allow the consumer to get a single object representing the line split at the configured break value. Node doesn't seem to have a way to create a buffer for which the underlying data is a list of buffers. Since the consumer expects a Buffer-like object, I didn't want to build such a wrapper. When I saw the performance was good enough for my use case, I left it at `concat`. If you know of a better way that doesn't require a ton of engineering, I'm open to suggestions.

Reply:
FYI I did find https://www.npmjs.com/package/buffers, but that basically does a concat when producing a slice out, so that's not any better than this.

Reply:
Does calling `this._writeItem` on each chunk not work? I would think that would avoid an intermediate allocation.

Reply:
By default node streams may concatenate bytes as it sees fit. They also have an optional object mode, which treats the data in an opaque manner (even if the data is a Buffer of bytes).

In this case the goal of the transform is to produce a sequence of lines in the form of a `Buffer` of bytes, and consumers expect to handle a line as a single `Buffer` object. As such the readable portion of the stream transform is configured in object mode.

Some of the handling in the consumer involves pipping into another stream and converting either the whole line or a subset to a string. While I could probably build a "Buffers" object that kept a list of the original buffers and supported those operations somehow, I figured one intermediate allocation wouldn't be the end of the world.